libevmasm/Assembly.cpp

```diff
@@ -74,6 +76,209 @@ unsigned Assembly::codeSize(unsigned subTagSize) const
 	}
 }
 
+void Assembly::importAssemblyItemsFromJSON(Json::Value const& _code)
+{
+	solAssert(m_items.empty());
+	solRequire(_code.isArray(), AssemblyImportException, "Supplied JSON is not an array.");
+	for (auto current = begin(_code); current != end(_code); ++current)
+	{
+		auto const& item = m_items.emplace_back(createAssemblyItemFromJSON(*current, m_sourceList));
+		if (item == Instruction::JUMPDEST)
+			solThrow(AssemblyImportException, "JUMPDEST instruction without a tag");
+		else if (item.type() == AssemblyItemType::Tag)
+			solRequire(
+				++current != end(_code) &&
+					createAssemblyItemFromJSON(*current, m_sourceList) == Instruction::JUMPDEST,
+				AssemblyImportException,
+				"JUMPDEST expected after tag."
+			);
+	}
+}
+
+AssemblyItem Assembly::createAssemblyItemFromJSON(Json::Value const& _json, std::vector<std::string> const& _sourceList)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static set<string> const validMembers{"name", "begin", "end", "source", "value", "modifierDepth", "jumpType"};
+	for (auto const& member: _json.getMemberNames())
+		solRequire(
+			validMembers.count(member),
+			AssemblyImportException,
+			"Unknown member '" + member + "'. Valid members are " +
+			solidity::util::joinHumanReadable(validMembers, ", ") + "."
+		);
+	solRequire(isOfType<string>(_json["name"]), AssemblyImportException, "Member 'name' not of type string.");
+	solRequire(isOfTypeIfExists<int>(_json, "begin"), AssemblyImportException, "Optional member 'begin' not of type int.");
+	solRequire(isOfTypeIfExists<int>(_json, "end"), AssemblyImportException, "Optional member 'end' not of type int.");
+	solRequire(isOfTypeIfExists<int>(_json, "source"), AssemblyImportException, "Optional member 'source' not of type int.");
+	solRequire(isOfTypeIfExists<string>(_json, "value"), AssemblyImportException, "Optional member 'value' not of type string.");
+	solRequire(
+		isOfTypeIfExists<int>(_json, "modifierDepth"),
+		AssemblyImportException,
+		"Optional member 'modifierDepth' not of type int."
+	);
+	solRequire(
+		isOfTypeIfExists<string>(_json, "jumpType"),
+		AssemblyImportException,
+		"Optional member 'jumpType' not of type string."
+	);
+
+	string name = getOrDefault<string>(_json["name"], "");
+	solRequire(!name.empty(), AssemblyImportException, "Member 'name' was empty.");
+
+	SourceLocation location;
+	location.start = get<int>(_json["begin"]);
+	location.end = get<int>(_json["end"]);
+	int srcIndex = getOrDefault<int>(_json["source"], -1);
```

Comment:
The size of `int` is machine-dependent. And our emscripten builds are 32-bit so `int` could be a 16-bit value there (though it's not 32-bit in all aspects, so it could also be 32-bit - we'd need to check that in any case).

We should be using types like `int32_t` or `int64_t` for values like this.

Reply:
But just to clarify, I would not go chasing this problem too deep. I'd only deal with it in the new code you have added. If we're using `int`s somewhere down the stack, I'd leave that part as is, and some assertions and report it as a bug to deal with later.

Reply:
I totally agree. However, I guess I just used `int` here, because `start` and `end` are also `int`s here.. but yeah, I guess an overflow for `start` and `end` is even much more likely (on 32 bit machines) ;) hmm.. but that's basically a very good point.. maybe we should consider to just replace all usages of platform dependent `ints`.. @cameel @ekpyron I did not check yet, maybe there was already an issue for that.